### DISCUSSION TITLE
Having a problem with ImGui_ImplWin32_WndProcHandler

### DISCUSSION THREAD

**Discussion by ghost** [ocornut/imgui#7142] [New Users Build/Link/Run issues ONLY!]
The error:
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error	C3861	'ImGui_ImplWin32_WndProcHandler': identifier not found	imguikey	C:\Users\workspace\Documents\Visual Studio 2022\Projects\imguikey\main.cpp	97		

so when I try to build it, im having that error i've been trying to fix it for the past 4 days and still have no luck. Can someone please help me fix it? this might be a dumb request but please bear with me as i'm trying to learn. the part that is having the error is this

LRESULT CALLBACK WndProc(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam) {
    if (ImGui_ImplWin32_WndProcHandler(hWnd, msg, wParam, lParam))
        return true;

    ImGuiIO& io = ImGui::GetIO();

    switch (msg) {
    case WM_SIZE:
        if (wParam != SIZE_MINIMIZED) {
            d3dpp.BackBufferWidth = LOWORD(lParam);
            d3dpp.BackBufferHeight = HIWORD(lParam);
            d3dDevice->Reset(&d3dpp);
        }
        return 0;
    case WM_SYSCOMMAND:
        if ((wParam & 0xfff0) == SC_KEYMENU)
            return 0;
        break;
    case WM_DESTROY:
        PostQuitMessage(0);
        return 0;
    case WM_CLOSE:
        PostQuitMessage(0);
        return 0;
    case WM_KEYDOWN:
        if (wParam == VK_ESCAPE)
            PostQuitMessage(0);
        return 0;
    }

    return DefWindowProc(hWnd, msg, wParam, lParam);
}

**Comment by ocornut** [comment]
The `ImGui_ImplWin32_WndProcHandler` function needs to be forward declared in your code.
We don't declare it in `imgui_impl_win32.h` for subtle technical reasons.

See `imgui_impl_win32.h`: 
```cpp
// Win32 message handler your application need to call.
// - Intentionally commented out in a '#if 0' block to avoid dragging dependencies on <windows.h> from this helper.
// - You should COPY the line below into your .cpp code to forward declare the function and then you can call it.
// - Call from your application's message handler. Keep calling your message handler unless this function returns TRUE.

#if 0
extern IMGUI_IMPL_API LRESULT ImGui_ImplWin32_WndProcHandler(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam);
#endif
```


If you search for `ImGui_ImplWin32_WndProcHandler` in our repo, you will find 5 places where it is used, and in all 5 places the function is forward declared:

```cpp
// Forward declare message handler from imgui_impl_win32.cpp
extern IMGUI_IMPL_API LRESULT ImGui_ImplWin32_WndProcHandler(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam);

LRESULT WINAPI WndProc(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam)
{
    if (ImGui_ImplWin32_WndProcHandler(hWnd, msg, wParam, lParam))
        return true;
....
```

**Reply by ghost** [reply]
thanks alot man
